In [1]:
import os.path
import sys
import random
from operator import itemgetter
from collections import defaultdict
#----------------------------------------
#  Data input
#----------------------------------------

# Read a text file into a corpus (list of sentences (which in turn are lists of words))
# (taken from nested section of HW0)
def readFileToCorpus(f):
    """ Reads in the text file f which contains one sentence per line.
    """
    if os.path.isfile(f):
        file = open(f, "r") # open the input file in read-only mode
        i = 0 # this is just a counter to keep track of the sentence numbers
        corpus = [] # this will become a list of sentences
        print("Reading file ", f)
        for line in file:
            i += 1
            sentence = line.split() # split the line into a list of words
            #append this lis as an element to the list of sentences
            corpus.append(sentence)
            if i % 1000 == 0:
    	#print a status message: str(i) turns int i into a string
    	#so we can concatenate it
                sys.stderr.write("Reading sentence " + str(i) + "\n")
        #endif
    #endfor
        return corpus
    else:
    #ideally we would throw an exception here, but this will suffice
        print("Error: corpus file ", f, " does not exist")
        sys.exit() # exit the script
    #endif
#enddef


# Preprocess the corpus
def preprocess(corpus):
    #find all the rare words
    freqDict = defaultdict(int)
    for sen in corpus:
	    for word in sen:
	       freqDict[word] += 1
	#endfor
    #endfor

    #replace rare words with unk
    for sen in corpus:
        for i in range(0, len(sen)):
            word = sen[i]
            print(word)
            print(freqDict[word])
            if freqDict[word] < 2:

                sen[i] = UNK
	    #endif
	#endfor
    #endfor

    #bookend the sentences with start and end tokens
    for sen in corpus:
        sen.insert(0, start)
        sen.append(end)
    #endfor

    return corpus
#enddef

def preprocessTest(vocab, corpus):
    #replace test words that were unseen in the training with unk
    for sen in corpus:
        for i in range(0, len(sen)):
            word = sen[i]
            if word not in vocab:
                sen[i] = UNK
	    #endif
	#endfor
    #endfor

    #bookend the sentences with start and end tokens
    for sen in corpus:
        sen.insert(0, start)
        sen.append(end)
    #endfor

    return corpus
#enddef

# Constants
UNK = "UNK"     # Unknown word token
start = "<s>"   # Start-of-sentence token
end = "</s>"    # End-of-sentence-token


#--------------------------------------------------------------
# Language models and data structures
#--------------------------------------------------------------

# Parent class for the three language models you need to implement
class LanguageModel:
    # Initialize and train the model (ie, estimate the model's underlying probability
    # distribution from the training corpus)
    def __init__(self, corpus):
        print("""Your task is to implement four kinds of n-gram language models:
      a) an (unsmoothed) unigram model (UnigramModel)
      b) a unigram model smoothed using Laplace smoothing (SmoothedUnigramModel)
      c) an unsmoothed bigram model (BigramModel)
      d) a bigram model smoothed using linear interpolation smoothing (SmoothedBigramModelInt)
      """)
    #enddef

    # Generate a sentence by drawing words according to the
    # model's probability distribution
    # Note: think about how to set the length of the sentence
    #in a principled way
    def generateSentence(self):
        print("Implement the generateSentence method in each subclass")
        return "mary had a little lamb ."
    #emddef

    # Given a sentence (sen), return the probability of
    # that sentence under the model
    def getSentenceProbability(self, sen):
        print("Implement the getSentenceProbability method in each subclass")
        return 0.0
    #enddef

    # Given a corpus, calculate and return its perplexity
    #(normalized inverse log probability)
    def getCorpusPerplexity(self, corpus):
        print("Implement the getCorpusPerplexity method")
        return 0.0
    #enddef

    # Given a file (filename) and the number of sentences, generate a list
    # of sentences and write each to file along with its model probability.
    # Note: you shouldn't need to change this method
    def generateSentencesToFile(self, numberOfSentences, filename):
        filePointer = open(filename, 'w+')
        for i in range(0,numberOfSentences):
            sen = self.generateSentence()
            prob = self.getSentenceProbability(sen)

            stringGenerated = str(prob) + " " + " ".join(sen)
            print(stringGenerated, end="\n", file=filePointer)

	#endfor
    #enddef
#endclass

# Unigram language model
class UnigramModel(LanguageModel):
    def __init__(self, corpus):
        print("Subtask: implement the unsmoothed unigram language model")
    #endddef
#endclass

#Smoothed unigram language model (use laplace for smoothing)
class SmoothedUnigramModel(LanguageModel):
    def __init__(self, corpus):
        print("Subtask: implement the smoothed unigram language model")
    #endddef
#endclass

# Unsmoothed bigram language model
class BigramModel(LanguageModel):
    def __init__(self, corpus):
        print("Subtask: implement the unsmoothed bigram language model")
    #endddef
#endclass



import os.path
import sys
import random
from operator import itemgetter
from collections import defaultdict, Counter
import math

#----------------------------------------
# Data input
#----------------------------------------

# Read a text file into a corpus (list of sentences (which in turn are lists of words))
def readFileToCorpus(f):
    """ Reads in the text file f which contains one sentence per line. """
    if os.path.isfile(f):
        file = open(f, "r")  # open the input file in read-only mode
        i = 0  # this is just a counter to keep track of the sentence numbers
        corpus = []  # this will become a list of sentences
        print("Reading file ", f)
        for line in file:
            i += 1
            sentence = line.split()  # split the line into a list of words
            corpus.append(sentence)
            if i % 1000 == 0:
                sys.stderr.write("Reading sentence " + str(i) + "\n")
            # endfor
        return corpus
    else:
        print("Error: corpus file ", f, " does not exist")
        sys.exit()  # exit the script
    # endif

# Preprocess the corpus
def preprocess(corpus):
    freqDict = defaultdict(int)
    for sen in corpus:
        for word in sen:
            freqDict[word] += 1
    for sen in corpus:
        for i in range(0, len(sen)):
            word = sen[i]
            if freqDict[word] < 2:
                sen[i] = UNK
        sen.insert(0, start)
        sen.append(end)
    return corpus

# Constants
UNK = "UNK"  # Unknown word token
start = "<s>"  # Start-of-sentence token
end = "</s>"  # End-of-sentence-token

#--------------------------------------------------------------
# Language models and data structures
#--------------------------------------------------------------

# Parent class for the four language models you need to implement
class LanguageModel:
    # Initialize and train the model (ie, estimate the model's underlying probability
    # distribution from the training corpus)
    def __init__(self, corpus):
        self.corpus = preprocess(corpus)
        self.vocab = set([word for sen in self.corpus for word in sen])
        self.vocab_size = len(self.vocab)
        self.unigram_counts = Counter([word for sentence in self.corpus for word in sentence])
        self.bigram_counts = Counter([(sentence[i], sentence[i+1]) for sentence in self.corpus for i in range(len(sentence)-1)])
        self.total_words = sum(self.unigram_counts.values())
        self.total_bigrams = sum(self.bigram_counts.values())

    # Generate a sentence by drawing words according to the
    # model's probability distribution
    # Note: think about how to set the length of the sentence
    #in a principled way
    def generateSentence(self):
        sentence = [start]
        while sentence[-1] != end:
            word = self.drawWord(sentence)
            sentence.append(word)
        return sentence

    # Given a sentence (sen), return the probability of
    # that sentence under the model
    def getSentenceProbability(self, sen):
        probability = 1
        for i in range(len(sen)-1):
            bigram_count = self.bigram_counts[(sen[i], sen[i+1])]
            unigram_count = self.unigram_counts[sen[i]]
            if bigram_count > 0:
                probability *= bigram_count / unigram_count
            else:
                probability *= 0.4 * unigram_count / self.total_words
        return probability

    # Given a corpus, calculate and return its perplexity
    #(normalized inverse log probability)
    def getCorpusPerplexity(self, corpus):
        corpus = preprocessTest(self.vocab, corpus)
        log_sum = 0
        total_words = 0
        for sentence in corpus:
            total_words += len(sentence) - 1
            log_prob = math.log(self.getSentenceProbability(sentence))
            log_sum += log_prob
        perplexity = math.exp(-log_sum / total_words)
        return perplexity

    # Generate a single random word according to the distribution
    def drawWord(self, sentence):
        rand = random.random()
        prob_sum = 0
        for word in self.vocab:
            prob_sum += self.getBigramProbability(sentence[-1], word)
            if rand <= prob_sum:
                return word
        return end

    # Given two words, return the probability of the second word
    # given the first word
    def getBigramProbability(self, word1, word2):
        bigram_count = self.bigram_counts[(word1, word2)]
        unigram_count = self.unigram_counts[word1]
        if bigram_count > 0:
            return bigram_count / unigram_count
        else:
            return 0.4 * unigram_count / self.total_words

# Unigram language model
class UnigramModel(LanguageModel):
    def __init__(self, corpus):
        super().__init__(corpus)

    def generateSentence(self):
        sentence = [start]
        while sentence[-1] != end:
            word = self.drawWord(sentence)
            sentence.append(word)
        return sentence

    def getSentenceProbability(self, sen):
        probability = 1
        for i in range(len(sen)-1):
            unigram_count = self.unigram_counts[sen[i]]
            probability *= unigram_count / self.total_words
        return probability

    def generateSentencesToFile(self, numberOfSentences, filename):
        filePointer = open(filename, 'w+')
        for i in range(numberOfSentences):
            sen = self.generateSentence()
            prob = self.getSentenceProbability(sen)

            stringGenerated = str(prob) + " " + " ".join(sen)
            print(stringGenerated, end="\n", file=filePointer)
        filePointer.close()

    def drawWord(self, sentence):
        rand = random.random()
        prob_sum = 0
        for word in self.vocab:
            prob_sum += self.unigram_counts[word] / self.total_words
            if rand <= prob_sum:
                return word
        return end

# Smoothed unigram language model (use laplace for smoothing)
class SmoothedUnigramModel(LanguageModel):
    def __init__(self, corpus):
        super().__init__(corpus)

    def generateSentence(self):
        sentence = [start]
        while sentence[-1] != end:
            word = self.drawWord(sentence)
            sentence.append(word)
        return sentence

    def getSentenceProbability(self, sen):
        probability = 1
        for i in range(len(sen)-1):
            unigram_count = self.unigram_counts[sen[i]]
            probability *= (unigram_count + 1) / (self.total_words + self.vocab_size)
        return probability

    def generateSentencesToFile(self, numberOfSentences, filename):
        filePointer = open(filename, 'w+')
        for i in range(numberOfSentences):
            sen = self.generateSentence()
            prob = self.getSentenceProbability(sen)

            stringGenerated = str(prob) + " " + " ".join(sen)
            print(stringGenerated, end="\n", file=filePointer)
        filePointer.close()

    def drawWord(self, sentence):
        rand = random.random()
        prob_sum = 0
        for word in self.vocab:
            prob_sum += (self.unigram_counts[word] + 1) / (self.total_words + self.vocab_size)
            if rand <= prob_sum:
                return word
        return end

# Unsmoothed bigram language model
class BigramModel(LanguageModel):
    def __init__(self, corpus):
        super().__init__(corpus)

    def generateSentence(self):
        sentence = [start]
        while sentence[-1] != end:
            word = self.drawWord(sentence)
            sentence.append(word)
        return sentence

    def getSentenceProbability(self, sen):
        probability = 1
        for i in range(len(sen)-1):
            bigram_count = self.bigram_counts[(sen[i], sen[i+1])]
            unigram_count = self.unigram_counts[sen[i]]
            if bigram_count > 0:
                probability *= bigram_count / unigram_count
            else:
                probability *= 0.4 * unigram_count / self.total_words
        return probability

    def generateSentencesToFile(self, numberOfSentences, filename):
        filePointer = open(filename, 'w+')
        for i in range(numberOfSentences):
            sen = self.generateSentence()
            prob = self.getSentenceProbability(sen)

            stringGenerated = str(prob) + " " + " ".join(sen)
            print(stringGenerated, end="\n", file=filePointer)
        filePointer.close()

    def drawWord(self, sentence):
        rand = random.random()
        prob_sum = 0
        for word in self.vocab:
            prob_sum += self.getBigramProbability(sentence[-1], word)
            if rand <= prob_sum:
                return word
        return end

# Smoothed bigram language model (use linear interpolation for smoothing, set lambda1 = lambda2 = 0.5)
class SmoothedBigramModelInt(LanguageModel):
    def __init__(self, corpus):
        super().__init__(corpus)
        self.lambda1 = 0.5
        self.lambda2 = 0.5

    def generateSentence(self):
        sentence = [start]
        while sentence[-1] != end:
            word = self.drawWord(sentence)
            sentence.append(word)
        return sentence

    def getSentenceProbability(self, sen):
        probability = 1
        for i in range(len(sen)-1):
            probability *= self.getSmoothedBigramProbability(sen[i], sen[i+1])
        return probability

    def generateSentencesToFile(self, numberOfSentences, filename):
        filePointer = open(filename, 'w+')
        for i in range(numberOfSentences):
            sen = self.generateSentence()
            prob = self.getSentenceProbability(sen)

            stringGenerated = str(prob) + " " + " ".join(sen)
            print(stringGenerated, end="\n", file=filePointer)
        filePointer.close()

    def drawWord(self, sentence):
        rand = random.random()
        prob_sum = 0
        for word in self.vocab:
            prob_sum += self.getSmoothedBigramProbability(sentence[-1], word)
            if rand <= prob_sum:
                return word
        return end

    def getSmoothedBigramProbability(self, word1, word2):
        bigram_count = self.bigram_counts[(word1, word2)]
        unigram_count = self.unigram_counts[word1]
        unigram_total = sum(self.unigram_counts.values())

        probability = (self.lambda1 * (bigram_count / unigram_count)) + (self.lambda2 * (unigram_count / unigram_total))
        return probability


# Sample class for a unsmoothed unigram probability distribution
# Note:
#       Feel free to use/re-use/modify this class as necessary for your
#       own code (e.g. converting to log probabilities after training).
#       This class is intended to help you get started
#       with your implementation of the language models above.
class UnigramDist:
    def __init__(self, corpus):
        self.counts = defaultdict(float)
        self.total = 0.0
        self.train(corpus)
    #endddef

    # Add observed counts from corpus to the distribution
    def train(self, corpus):
        for sen in corpus:
            for word in sen:
                if word == start:
                    continue
                self.counts[word] += 1.0
                self.total += 1.0
            #endfor
        #endfor
    #enddef

    # Returns the probability of word in the distribution
    def prob(self, word):
        return self.counts[word]/self.total
    #enddef

    # Generate a single random word according to the distribution
    def draw(self):
        rand = random.random()
        for word in self.counts.keys():
            rand -= self.prob(word)
            if rand <= 0.0:
                return word
	    #endif
	#endfor
    #enddef
#endclass

#-------------------------------------------
# The main routine
#-------------------------------------------
if __name__ == "__main__":
    trainCorpus = readFileToCorpus('train.txt')
    trainCorpus = preprocess(trainCorpus)

    posTestCorpus = readFileToCorpus('pos_test.txt')
    negTestCorpus = readFileToCorpus('neg_test.txt')

    vocab = set()
    for sentence in trainCorpus:
        vocab.update(sentence)

    posTestCorpus = preprocessTest(vocab, posTestCorpus)
    negTestCorpus = preprocessTest(vocab, negTestCorpus)

    # Creating language models
    unigram_model = UnigramModel(trainCorpus)
    smoothed_unigram_model = SmoothedUnigramModel(trainCorpus)
    bigram_model = BigramModel(trainCorpus)
    smoothed_bigram_model = SmoothedBigramModelInt(trainCorpus)

    # Generating sentences and writing to files
    unigram_model.generateSentencesToFile(20, 'unigram_output.txt')
    smoothed_unigram_model.generateSentencesToFile(20, 'smooth_unigram_output.txt')
    bigram_model.generateSentencesToFile(20, 'bigram_output.txt')
    smoothed_bigram_model.generateSentencesToFile(20, 'smooth_bigram_output.txt')

    # Calculating perplexity
    unigram_perplexity_pos = unigram_model.getCorpusPerplexity(posTestCorpus)
    unigram_perplexity_neg = unigram_model.getCorpusPerplexity(negTestCorpus)

    smoothed_unigram_perplexity_pos = smoothed_unigram_model.getCorpusPerplexity(posTestCorpus)
    smoothed_unigram_perplexity_neg = smoothed_unigram_model.getCorpusPerplexity(negTestCorpus)

    bigram_perplexity_pos = bigram_model.getCorpusPerplexity(posTestCorpus)
    bigram_perplexity_neg = bigram_model.getCorpusPerplexity(negTestCorpus)

    smoothed_bigram_perplexity_pos = smoothed_bigram_model.getCorpusPerplexity(posTestCorpus)
    smoothed_bigram_perplexity_neg = smoothed_bigram_model.getCorpusPerplexity(negTestCorpus)

    print(" ")
    print("UNIGRAM : ")
    print("------------------------------------------------------------------")
    print("Unigram Perplexity (Positive Test Corpus):", unigram_perplexity_pos)
    print("Unigram Perplexity (Negative Test Corpus):", unigram_perplexity_neg)
    print(" ")
    print("SMOOTHED UNIGRAM  : ")
    print("------------------------------------------------------------------")
    print("Smoothed Unigram Perplexity (Positive Test Corpus):", smoothed_unigram_perplexity_pos)
    print("Smoothed Unigram Perplexity (Negative Test Corpus):", smoothed_unigram_perplexity_neg)
    print(" ")
    print("BIGRAM : ")
    print("------------------------------------------------------------------")
    print("Bigram Perplexity (Positive Test Corpus):", bigram_perplexity_pos)
    print("Bigram Perplexity (Negative Test Corpus):", bigram_perplexity_neg)
    print(" ")
    print("SMOOTHED BIGRAM : ")
    print("------------------------------------------------------------------")
    print("Smoothed Bigram Perplexity (Positive Test Corpus):", smoothed_bigram_perplexity_pos)
    print("Smoothed Bigram Perplexity (Negative Test Corpus):", smoothed_bigram_perplexity_neg)




Reading file  train.txt


Reading sentence 1000
Reading sentence 2000
Reading sentence 3000
Reading sentence 4000
Reading sentence 5000
Reading sentence 6000
Reading sentence 7000
Reading sentence 8000
Reading sentence 9000
Reading sentence 10000
Reading sentence 11000
Reading sentence 12000
Reading sentence 13000
Reading sentence 14000
Reading sentence 15000
Reading sentence 16000
Reading sentence 17000
Reading sentence 18000
Reading sentence 19000
Reading sentence 20000
Reading sentence 21000
Reading sentence 22000
Reading sentence 23000
Reading sentence 24000
Reading sentence 25000
Reading sentence 26000
Reading sentence 27000


Reading file  pos_test.txt
Reading file  neg_test.txt


Reading sentence 28000
Reading sentence 29000
Reading sentence 30000
Reading sentence 1000
Reading sentence 1000


 
UNIGRAM : 
------------------------------------------------------------------
Unigram Perplexity (Positive Test Corpus): 516.6333952693551
Unigram Perplexity (Negative Test Corpus): 497.90085561478963
 
SMOOTHED UNIGRAM  : 
------------------------------------------------------------------
Smoothed Unigram Perplexity (Positive Test Corpus): 406.1064967347534
Smoothed Unigram Perplexity (Negative Test Corpus): 388.0202224857519
 
BIGRAM : 
------------------------------------------------------------------
Bigram Perplexity (Positive Test Corpus): 79.81613647736447
Bigram Perplexity (Negative Test Corpus): 80.69852371189012
 
SMOOTHED BIGRAM : 
------------------------------------------------------------------
Smoothed Bigram Perplexity (Positive Test Corpus): 52.1718508581214
Smoothed Bigram Perplexity (Negative Test Corpus): 53.389779919118105


Questions

**Q1**

In [ ]:
# Unigram predicts the probability of each indivual words where as bigrams predict the probability of the next word based on the previous word

**Q2**

In [ ]:
# Certainly! The models are designed to assign notably different probabilities to sentences. Unigram and smoothed unigram models may favor sentences with frequently occurring words,
# whereas bigram and smoothed bigram models take into account word dependencies, leading to diverse probability distributions.

**Q3**

In [ ]:
# bigram models are more effective as they add a touch of accuracy and is more likely to suggest better sentence creation.

**Q4**

**Unigram Model:**

Positive Test Corpus: 516.63

Negative Test Corpus: 497.90

**Smoothed Unigram Model (Laplace Smoothing):**

Positive Test Corpus: 406.11

Negative Test Corpus: 388.02

**Bigram Model:**

Positive Test Corpus: 79.82

Negative Test Corpus: 80.70

**Smoothed Bigram Model (Linear Interpolation Smoothing):**

Positive Test Corpus: 52.17

Negative Test Corpus: 53.39

These perplexity values quantify how well each model aligns with the respective test corpora. Lower perplexity values indicate better performance, suggesting that the model is more adept at predicting the given data.